src/db.rs

```diff
@@ -260,43 +261,60 @@ impl DbInner {
         }
     }
 
-    async fn replay_wal(&self) -> Result<(), SlateDBError> {
+    async fn replay_wal(self: &Arc<Self>) -> Result<(), SlateDBError> {
         let wal_id_last_compacted = self.state.read().state().core.last_compacted_wal_sst_id;
         let wal_sst_list = self
             .table_store
             .get_wal_sst_list(wal_id_last_compacted)
             .await?;
         let mut last_sst_id = wal_id_last_compacted;
-        for sst_id in wal_sst_list {
-            last_sst_id = sst_id;
-            let sst = self.table_store.open_sst(&SsTableId::Wal(sst_id)).await?;
-            let sst_id = match &sst.id {
-                SsTableId::Wal(id) => *id,
-                SsTableId::Compacted(_) => return Err(SlateDBError::InvalidDBState),
-            };
-            let mut iter = SstIterator::new(&sst, self.table_store.clone(), 1, 1, true).await?;
-            // iterate over the WAL SSTs in reverse order to ensure we recover in write-order
-            // buffer the WAL entries to bulk replay them into the memtable.
-            let mut wal_replay_buf = Vec::new();
-            while let Some(kv) = iter.next_entry().await? {
-                wal_replay_buf.push(kv);
+
+        for sst_chunk in wal_sst_list.chunks(4) {
+            let mut wal_tasks = Vec::new();
+            let mut wal_replay_errors = Vec::new();
+            // constructs a list of tokio tasks to asynchronously fetch wal sst
+            for &sst_id in sst_chunk {
+                let this = self.clone();
+                wal_tasks.push(tokio::spawn(async move { this.fetch_wal(sst_id).await }));
```

Comment:
This is what I was trying to avoid with my previous suggestion (https://github.com/slatedb/slatedb/pull/210/files#r1759599316). Introducing Tokio tasks does increase parallelism, but at the cost of implementation complexity. What I was suggesting in my previous post was that we simply maintain a VecDeque<SstIterator>. We seed this vec with iterators for the first 4 WAL SSTs (using new_spawn with 1 fetch task and 256 blocks). We then pop an SstIterator off the top of the vector and iterate over it as usual. After the iterator is exhausted, we push a _new_ SstIterator into the vec again.

The approach I'm proposing means don't need any tokio async tasks. Because we're new_spawn'ing the 4 SstIterators start their fetches in parallel immediately upon instantiation. When we finish one iterator and push a new one on to the talk of the queue, it again begins fetching immediately. This gives us some parallelism without having to add any new tokio tasks/futures to the WAL replay code.

Could you use this strategy instead of the tokio-based approach?

Reply:
ah alright, the previous comments confused me a bit, i thought they suggested combining the tokio::tasks with new_spawn. I will try to make these changes today